clang/test/CodeGenSYCL/virtual-types.cpp

```diff
@@ -21,8 +21,8 @@ int main() {
 // CHECK: %struct.Struct = type { ptr }
 // VTable:
 // CHECK: @_ZTV6Struct = linkonce_odr unnamed_addr constant { [3 x ptr] } { [3 x ptr] [ptr null, ptr @_ZTI6Struct, ptr @_ZN6Struct3fooEv] }, comdat, align 8
-// CHECK: @[[TYPEINFO:.+]] = external addrspace(1) global ptr addrspace(4)
+// CHECK: @[[TYPEINFO:.+]] = external global ptr 
```

Comment:
I THINK we intended to have generic address space here, i.e addrspace (4), but I'm not sure.
CC @KornevNikita @AlexeySachkov for opinion.

Reply:
I only changed it because it failed after my clang patch, so I guess it can be switched to opaques if it passes now

Reply:
The question is not whether we should use opaque pointers in the test, but which address space should be here.

Reply:
Yeah, now we seem to have a private pointer to a private pointer (double indirection, because every global is a pointer by itself). Having them both private is a problem, which we previously addressed in https://github.com/intel/llvm/pull/7449

It is not yet clear to me why exactly and how those address spaces were lost, but they are important. I don't remember which one of them is the most crucial here, but without it we emit invalid SPIR-V and can't reverse-translate it back later (I've just tested that locally)

Reply:
Incorrect address space was introduced here: 66c18fee337fdf04f04f48e0fefb68157915b8a8

```
commit 66c18fee337fdf04f04f48e0fefb68157915b8a8
Author: Tyler Lanphear <tylanphear@gmail.com>
Date:   Mon Aug 14 19:02:51 2023 -0700

    [sycl-web] Fit LIT tests failures in CodeGenSYCL. (#10820)

    Fix a number of LIT test failures by reverting some changes,
    and re-adding some others behind a guard.

    - Revert "[Clang][CodeGen]`vtable`, `typeinfo` et al. are globals"
    - Guard opaque ptr changes behind INTEL_SYCL_OPAQUEPOINTER_READY.
```

In particular, here is the change:

```diff
-    VTable =
-        CGM.getModule().getOrInsertGlobal(VTableName, CGM.GlobalsInt8PtrTy);
+#ifdef INTEL_SYCL_OPAQUEPOINTER_READY
+    VTable = CGM.getModule().getOrInsertGlobal(VTableName, CGM.Int8PtrTy);
+#else  // INTEL_SYCL_OPAQUEPOINTER_READY
+    VTable = CGM.CreateRuntimeVariable(CGM.DefaultInt8PtrTy, VTableName);
+#endif // INTEL_SYCL_OPAQUEPOINTER_READY
```

As you can see, in the upstream we use `GlobalsInt8PtrTy`, which is wrong for SPIR-V targets just yet (because function pointers are still emitted in private and you can't have private -> global casts). In our version we have used `DefaultInt8PtrTy`, which is essentially a generic and under `opaquepointer ready` macro we put `Int8PtrTy`, i.e. "default" address space.

Reply:
Thanks @AlexeySachkov !

With above fix, we still have that missing, so there maybe other instances of similar change?

@_ZTI6Struct = linkonce_odr constant { ptr, ptr } { ptr getelementptr inbounds (ptr, ptr @_ZTVN10__cxxabiv117__class_type_infoE, i64 2), ptr @_ZTS6Struct }, comdat, align 8

Reply:
Other changes from that commit in the same file also look incorrect:

```diff
         llvm::ConstantExpr::getInBoundsGetElementPtr(CGM.Int8Ty, VTable, Eight);
   } else {
     llvm::Constant *Two = llvm::ConstantInt::get(PtrDiffTy, 2);
-    VTable = llvm::ConstantExpr::getInBoundsGetElementPtr(CGM.GlobalsInt8PtrTy,
+#ifdef INTEL_SYCL_OPAQUEPOINTER_READY
+    VTable = llvm::ConstantExpr::getInBoundsGetElementPtr(CGM.Int8PtrTy,
+#else  // INTEL_SYCL_OPAQUEPOINTER_READY
+    VTable = llvm::ConstantExpr::getInBoundsGetElementPtr(CGM.DefaultInt8PtrTy,
+#endif // INTEL_SYCL_OPAQUEPOINTER_READY
                                                           VTable, Two);
   }
+#ifdef INTEL_SYCL_OPAQUEPOINTER_READY
   VTable = llvm::ConstantExpr::getBitCast(VTable, CGM.GlobalsInt8PtrTy);
+#else  // INTEL_SYCL_OPAQUEPOINTER_READY
+  VTable = llvm::ConstantExpr::getBitCast(VTable, CGM.DefaultInt8PtrTy);
+#endif // INTEL_SYCL_OPAQUEPOINTER_READY
```

Once again, `GlobalsInt8PtrTy` should _not_ be used and its removal is expected as an effect of a community patch revert.
However, Change to `Int8PtrTy` shouldn't have been done, we should continue to use `DefaultInt8PtrTy` as we used to in `sycl` branch

Reply:
Yes, but simply changing all the `CGM.Int8PtrTy` to `CGM.DefaultInt8PtrTy` will lead to an assert.

Reply:
Update the branch to avoid using `GlobalsInt8PtrTy`, however, this will lead to assert.

Reply:
Let me try that myself as well, I have `sycl-web` built locally